src/orange/orangeinp/CsgTreeUtils.cc

```diff
@@ -165,6 +171,273 @@ void simplify(CsgTree* tree, NodeId start)
     }
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Simplify negated joins using De Morgan's law.
+ *
+ * This is required if the tree's logic expression is used with
+ * \c InfixEvaluator as negated joins are not supported.
+ */
+CsgTree transform_negated_joins(CsgTree const& tree)
```

Comment:
This function is too long; is there a sensible way to break it up into a detail helper class? Probably making these maps as class data, maybe a couple helper functions that you could call in sequence, turn the lambda into a member function.